### PR TITLE
Longitudinal aperture

### PR DESCRIPTION
**Pull request by SebastienJoly**
Introduction of the LongAperturePass integrator in atintegrators and LongAperture element to add the possibility to use rectangular longitudinal apertures.

Small test using the longitudinal acceptance module below:
First without any longitudinal aperture
image
After adding longitudinal apertures at ct = $\pm$ 0.1 
image

I duplicated the transverse aperture functions/modules to create their longitudinal counterparts. Should we stick to this option or have general functions which can do both planes depending on specific arguments?

Moreover, new tests for the LongAperture element would be a great idea in the future.

### CODE REVIEW

**Comment by swhite2401** [comment]
Hi @SebastienJoly is this ready for review?

**Comment by SebastienJoly** [comment]
Hello @swhite2401 
Yes this is ready for review!

File: atintegrators/AperturePass.c [hunk 1]
```diff
@@ -65,7 +67,7 @@ void mexFunction(	int nlhs, mxArray *plhs[], int nrhs, const mxArray *prhs[])
         }
     }
     else {
-        mexErrMsgIdAndTxt("AT:WrongArg","Needs 0 or 2 arguments");
+        mexErrMsgIdAndTxt("AT:WrongArg","Needs 0, 2 or 3 arguments");
```
Comment (swhite2401) [inline atintegrators/AperturePass.c:70]: Why did you change this field? nrhs == 3 is not implemented....

**Review by swhite2401 — COMMENTED** [review]
Besides this change in help in the MATLAB mexfunction everything looks fine to me

File: atintegrators/AperturePass.c (hunk 1 above)
Comment (lfarv) [reply atintegrators/AperturePass.c:70]: To avoid confusion, I would even keep `AperturePass.c` unchanged. This was probably an attempt to create a general function for the 3 planes, but I do prefer the present option to have a separate element and integrators for the longitudinal aperture.

**Comment by lfarv** [comment]
In `elements.py` there is a `LongtMotion` class (unrelated to this PR). I initially called it `LongMotion`, and I got the remark that the shortcut "Long" for "Longitudinal" was ambiguous: the class deals with "longitudinal motion", not "long motion".

The same applies here: we are not talking of a "long aperture". So I propose a `LongtAperture` element, and a `LongtAperturePass.c` integrator.

Oherwise, this looks OK.

File: atintegrators/AperturePass.c (hunk 1 above)
Comment (SebastienJoly) [reply atintegrators/AperturePass.c]: Indeed, my mistake. I did try to create a function for all three planes but decided to split it in two functions instead. I forgot to change the error message.

**Comment by SebastienJoly** [comment]
> In `elements.py` there is a `LongtMotion` class (unrelated to this PR). I initially called it `LongMotion`, and I got the remark that the shortcut "Long" for "Longitudinal" was ambiguous: the class deals with "longitudinal motion", not "long motion".
> 
> The same applies here: we are not talking of a "long aperture". So I propose a `LongtAperture` element, and a `LongtAperturePass.c` integrator.
> 
> Oherwise, this looks OK.

Good point, I changed the naming convention to LongtAperture. This is clearer and now consistent with the rest of the code as you said.

**Comment by SebastienJoly** [comment]
If there is no other comment, can I merge it?

**Comment by swhite2401** [comment]
I am not so sure but you can try to remove `LongAperturePass.c` to keep only `LongtAperturePass.c` and change the function name (and dependents) to  `LongtAperturePass` instead of `LongAperturePass` to be consistent with other pass methods

File: atintegrators/LongtAperturePass.c [hunk 2]
```diff
@@ -0,0 +1,73 @@
+
+#include "atelem.c"
+#include "atlalib.c"
+
+struct elem {
+    double *limits;
+};
+
+void LongAperturePass(double *r_in, double *limits, int num_particles)
```
Comment (swhite2401) [inline atintegrators/LongtAperturePass.c:9]: Here try `LongtAperturePass`

File: atintegrators/LongtAperturePass.c [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+#if defined(MATLAB_MEX_FILE) || defined(PYAT)
+ExportMode struct elem *trackFunction(const atElem *ElemData,struct elem *Elem,
+        double *r_in, int num_particles, struct parameters *Param)
+{
+    if (!Elem) {
+        double *limits=atGetDoubleArray(ElemData,"Limits"); check_error();
+        Elem = (struct elem*)atMalloc(sizeof(struct elem));
+        Elem->limits=limits;
+        }
+    LongAperturePass(r_in, Elem->limits, num_particles);
```
Comment (swhite2401) [inline atintegrators/LongtAperturePass.c:37]: `LongtAperturePass`

File: atintegrators/LongtAperturePass.c [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    if (nrhs == 2) {
+        double *r_in;
+        const mxArray *ElemData = prhs[0];
+        int num_particles = mxGetN(prhs[1]);
+        double *limits=atGetDoubleArray(ElemData,"Limits"); check_error();
+        if (mxGetM(prhs[1]) != 6) mexErrMsgIdAndTxt("AT:WrongArg","Second argument must be a 6 x N matrix");
+        /* ALLOCATE memory for the output array of the same size as the input  */
+        plhs[0] = mxDuplicateArray(prhs[1]);
+        r_in = mxGetDoubles(plhs[0]);
+        LongAperturePass(r_in,limits, num_particles);
```
Comment (swhite2401) [inline atintegrators/LongtAperturePass.c:58]: `LongtAperturePass`

File: atintegrators/atlalib.c [hunk 5]
```diff
@@ -170,9 +170,18 @@ static void checkiflostRectangularAp(double *r6, const double *limits)
 {
 	/* check limits for X position */
     if (r6[0]<limits[0] || r6[0]>limits[1])      markaslost(r6,5);
+	/* check limits for Y position */
     else if (r6[2]<limits[2] || r6[2]>limits[3]) markaslost(r6,5);
 }
 
+static void checkiflostLongRectangularAp(double *r6, const double *limits)
```
Comment (swhite2401) [inline atintegrators/atlalib.c:177]: maybe use the same `Longt` convention...

File: pyat/at/lattice/elements.py [hunk 6]
```diff
@@ -266,6 +266,7 @@ class Element(object):
                         T1=lambda v: _array(v, (6,)),
                         T2=lambda v: _array(v, (6,)),
                         RApertures=lambda v: _array(v, (4,)),
+                        LRApertures=lambda v: _array(v, (4,)),
```
Comment (swhite2401) [inline pyat/at/lattice/elements.py:269]: Is this needed if restrict the implementation to a new `LongtAperture` element?

File: atintegrators/LongtAperturePass.c [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+     * limits has 4 elements: (MinDP, MaxDP, MinCT, MaxCT)
+     */
+    int c;
+    double *r6;
+    for (c=0; c<num_particles; c++) {
+        r6 = r_in+c*6;
+        if (!atIsNaN(r6[0]))
+        {
+            /*  check if this particle is already marked as lost */
+            checkiflostLongRectangularAp(r6, limits);
```
Comment (swhite2401) [inline atintegrators/LongtAperturePass.c:23]: `Longt` convention

**Comment by swhite2401** [comment]
Alright with the correct name all tests pass, I wait for you to remove the commented out line in` elements.py` and I can approve

**Comment by lfarv** [comment]
I still don't understand why `Aperture.c` is modified. Can you revert to the original file?

**Comment by SebastienJoly** [comment]
Yes I was waiting to see if the tests would finish successfully. Thanks a lot for your help!

File: atintegrators/AperturePass.c [hunk 8]
```diff
@@ -8,15 +8,16 @@ struct elem {
 
 void AperturePass(double *r_in, double *limits, int num_particles)
 {
-    /* Checks X and Y of each input 6-vector and marks the corresponding element in
-     * lossflag array with 0 if X,Y are exceed the limits given by limitsptr array
-     * limitsptr has 4 elements: (MinX, MaxX, MinY, MaxY)
+    /* Checks X,Y of each input 6-vector and marks the corresponding element in
+     * lossflag array with 0 if X,Y are exceed the limits given by limits and limitslong
+     * arrays
+     * limits has 4 elements: (MinX, MaxX, MinY, MaxY)
```
Comment (swhite2401) [inline atintegrators/AperturePass.c:14]: @lfarv I think this change is correct no? Why `limitsptr`? For the rest it is cosmetic, I have seen such changes in other PR in many occasions...

Comment (lfarv) [reply atintegrators/AperturePass.c:14]: What is `limitslong` ? But I agree with the change from `limitsptr` to `limits`.

Comment (SebastienJoly) [reply atintegrators/AperturePass.c:14]: I forgot to remove 'limitslong' from my previous try to handle both the longitudinal and transverse apertures in a single integrator. I corrected the docstrings in both integrators. There was also a mention of 'limitslong' in LongtAperturePass.c because I copy passed the docstring.

**Review by swhite2401 — APPROVED** [review]
ok for me

**Review by lfarv — APPROVED** [review]
OK
